rclrs/src/wait.rs

```diff
@@ -150,6 +166,38 @@ impl WaitSet {
         Ok(())
     }
 
+    /// Adds a guard condition to the wait set.
+    ///
+    /// # Errors
+    /// - If the guard condition was already added to this wait set or another one,
+    ///   [`AlreadyAddedToWaitSet`][1] will be returned
+    /// - If the number of guard condition in the wait set is larger than the
+    ///   capacity set in [`WaitSet::new`], [`WaitSetFull`][2] will be returned
+    ///
+    /// [1]: crate::RclrsError
+    /// [2]: crate::RclReturnCode
+    pub fn add_guard_condition(
+        &mut self,
+        guard_condition: Arc<Mutex<GuardCondition>>,
```

Comment:
Correction: I see now why you have wrapped it in a Mutex – because the `trigger()` method needs a mutable self.

The docs sound like the `rcl_trigger_guard_condition()` function can run concurrently with itself, and the "no" in thread-safe is only wrt the fini function: https://github.com/ros2/rcl/blob/4b125b1af0e2e2c8c7dd0c8e18b5a8d36709058c/rcl/include/rcl/guard_condition.h#L197

So, in theory, we could make `trigger()` take a shared reference and cast it to a mutable pointer (raw mutable pointers may alias). But it's a bit spicy.

If we don't do that, then I'd propose moving the `Mutex` inside the guard condition. The guard condition will always need to be accessed from different threads, so we can save the user some work there.